File: apps/api/src/smart_badge_api/schemas/recordings.py
from typing import Any, Literal

from pydantic import BaseModel, model_validator


class RecordingCreate(BaseModel):
    visit_id: str | None = None
    staff_id: str | None = None
    device_id: str | None = None


VALID_RECORDING_STATUSES = {"uploaded", "transcribing", "transcribed", "analyzing", "analyzed", "failed", "filtered"}


class RecordingUpdate(BaseModel):
    visit_id: str | None = None
    linked_visit_ids: list[str] | None = None
    staff_id: str | None = None
    device_id: str | None = None
    status: str | None = None
    duration_seconds: int | None = None

    @model_validator(mode="after")
    def _validate_status(self):
        if self.status is not None and self.status not in VALID_RECORDING_STATUSES:
            raise ValueError(f"status must be one of {VALID_RECORDING_STATUSES}")
        return self


class RecordingOut(BaseModel):
    class LinkedVisitOut(BaseModel):
        id: str
        external_visit_order_no: str | None = None
        external_visit_order_seg: str | None = None
        customer_name: str | None = None
        is_primary: bool = False

    id: str
    visit_id: str | None
    linked_visit_ids: list[str] = []
    linked_visits: list[LinkedVisitOut] = []
    visit_status: str | None = None
    staff_id: str | None
    staff_name: str | None = None
    staff_badge_id: str | None = None
    staff_role: str | None = None
    customer_name: str | None = None
    device_id: str | None
    device_code: str | None = None
    file_name: str
    file_size: int | None
    duration_seconds: int | None
    status: str
    split_parent_recording_id: str | None = None
    split_part_index: int | None = None
    split_at_ms: int | None = None
    has_transcript: bool = False
    created_at: str

    model_config = {"from_attributes": True}


class RecordingSplitRequest(BaseModel):
    split_at_seconds: float | None = None
    split_at_ms: int | None = None
    confirm: bool = False

    @model_validator(mode="after")
    def _validate_split_time(self):
        if self.split_at_seconds is None and self.split_at_ms is None:
            raise ValueError("split_at_seconds or split_at_ms is required")
        if self.split_at_seconds is not None and self.split_at_seconds <= 0:
            raise ValueError("split_at_seconds must be greater than 0")
        if self.split_at_ms is not None and self.split_at_ms <= 0:
            raise ValueError("split_at_ms must be greater than 0")
        return self

    def resolved_split_at_ms(self) -> int:
        if self.split_at_ms is not None:
            return int(self.split_at_ms)
        return int(round(float(self.split_at_seconds or 0) * 1000))


class RecordingSplitPartOut(BaseModel):
    part_index: int
    archive_item_id: str | None = None
    recording: RecordingOut


class RecordingSplitOut(BaseModel):
    original_recording_id: str
    split_at_ms: int
    parts: list[RecordingSplitPartOut]
    message: str


class ArchiveRecordingOut(BaseModel):
    id: str
    stage_key: str | None = None
    sn: str | None = None
    device_code: str | None = None
    file_id: str
    display_file_name: str
    archive_file_name: str | None = None
    staged_file_name: str | None = None
    remote_file_name: str | None = None
    audio_path: str | None = None
    archive_audio_path: str | None = None
    stage_audio_path: str | None = None
    duration_ms: int | None = None
    duration_seconds: int | None = None
    file_size: int | None = None
    create_time: str | None = None
    downloaded_at: str | None = None
    updated_at: str | None = None
    staff_id: str | None = None
    staff_name: str | None = None
    staff_role: str | None = None
    pipeline_status: str | None = None
    quality_stage: str | None = None
    quality_reason: str | None = None
    error_message: str | None = None
    recording_id: str | None = None
    is_split_hidden: bool = False
    visit_id: str | None = None
    linked_visit_ids: list[str] = []
    linked_visit_order_refs: list[str] = []
    linked_customer_names: list[str] = []
    has_visit_link: bool = False
    needs_visit_link: bool = False
    utterance_count: int | None = None
    full_text_length: int | None = None
    has_transcript: bool
    has_analysis: bool
    analysis_summary: dict[str, Any] | None = None


class ArchiveRecordingDateSummaryOut(BaseModel):
    date: str | None = None
    total: int = 0
    linked_count: int = 0
    needs_link_count: int = 0


class ArchiveRecordingPageOut(BaseModel):
    items: list[ArchiveRecordingOut]
    total: int
    page: int
    page_size: int
    pages: int
    date_summaries: list[ArchiveRecordingDateSummaryOut] = []


class ArchiveRecordingDetailOut(ArchiveRecordingOut):
    manifest: dict[str, Any] | None = None
    archive_metadata: dict[str, Any] | None = None
    transcript: dict[str, Any] | None = None
    analysis_result: dict[str, Any] | None = None
    analysis_summary: dict[str, Any] | None = None


class ArchiveRecordingEnsureOut(BaseModel):
    item_id: str
    recording_id: str
    file_name: str
    display_file_name: str
    created_new_recording: bool
    visit_id: str | None = None
    linked_visit_ids: list[str] = []
    linked_visit_order_refs: list[str] = []
    linked_customer_names: list[str] = []


class RecordingMediaSourceOut(BaseModel):
    url: str
    file_name: str
    media_type: str | None = None


class PendingArchiveRecordingOut(BaseModel):
    id: str
    display_file_name: str
    create_time: str | None = None
    duration_seconds: int | None = None
    staff_id: str | None = None
    staff_name: str | None = None
    device_code: str | None = None
    pipeline_status: str | None = None
    recording_id: str | None = None
    has_transcript: bool = False
    has_analysis: bool = False
    match_score: float = 0.0
    match_reasons: list[str] = []


class SapConsultationPayload(BaseModel):
    text: str
    user: str
    zxxx: dict[str, str]
    TAB_SYZ: list[dict[str, str]]


class SapPushTargetOut(BaseModel):
    visit_id: str | None = None
    visit_order_no: str
    visit_order_seg: str | None = None
    customer_name: str
    customer_code: str
    advisor_name: str
    indication_count: int
    recording_count: int = 1
    is_primary: bool = False


class SapPushPreviewOut(BaseModel):
    recording_id: str
    visit_order_no: str
    visit_order_seg: str | None = None
    customer_name: str
    customer_code: str
    advisor_name: str
    indication_count: int
    recording_count: int = 1
    target_count: int = 1
    targets: list[SapPushTargetOut] = []
    payloads: list[SapConsultationPayload]


class RecordingCustomerSegmentOut(BaseModel):
    id: str
    segment_index: int
    label: str
    begin_ms: int
    end_ms: int
    summary: str
    utterance_count: int
    status: str
    mapped_visit_id: str | None = None


class RecordingVisitAnalysisOut(BaseModel):
    id: str
    recording_id: str
    visit_id: str
    visit_order_no: str | None = None
    visit_order_seg: str | None = None
    customer_name: str | None = None
    customer_code: str | None = None
    customer_segment_id: str | None = None
    mapping_status: str
    analysis_status: str
    analysis_task_id: str | None = None
    analysis_error: str | None = None
    confirmed_by: str | None = None
    confirmed_at: str | None = None
    sap_ready_at: str | None = None
    sap_push_log_id: str | None = None


class RecordingMultiCustomerReviewOut(BaseModel):
    recording_id: str
    required: bool
    linked_visit_count: int
    status: Literal["not_required", "pending_mapping", "analyzing", "ready", "failed"]
    message: str
    segments: list[RecordingCustomerSegmentOut] = []
    visit_analyses: list[RecordingVisitAnalysisOut] = []


class RecordingMultiCustomerMappingIn(BaseModel):
    visit_id: str
    customer_segment_id: str


class RecordingMultiCustomerConfirmRequest(BaseModel):
    mappings: list[RecordingMultiCustomerMappingIn]


class SapPushDispatchRequest(BaseModel):
    trigger_mode: Literal["manual", "auto_bind", "scheduled"] = "manual"
    async_dispatch: bool = True
    target_visit_id: str | None = None


class SapPushAttemptOut(BaseModel):
    request_index: int
    success: bool
    http_status_code: int | None = None
    gateway_code: int | str | None = None
    business_status: str | None = None
    business_message: str | None = None
    response_body: Any = None


class SapPushLogOut(BaseModel):
    id: str
    recording_id: str | None = None
    recording_file_name: str | None = None
    recording_created_at: str | None = None
    visit_id: str | None = None
    visit_order_no: str | None = None
    visit_order_seg: str | None = None
    customer_name: str | None = None
    customer_code: str | None = None
    advisor_name: str | None = None
    trigger_mode: str
    status: str
    send_enabled: bool
    initiated_by: str | None = None
    request_url: str | None = None
    trace_id: str | None = None
    request_payloads: list[dict[str, Any]] = []
    gateway_requests: list[dict[str, Any]] = []
    response_items: list[SapPushAttemptOut] = []
    http_status_code: int | None = None
    business_status: str | None = None
    business_message: str | None = None
    error_message: str | None = None
    effective_status: str | None = None
    effective_business_status: str | None = None
    effective_reason: str | None = None
    sent_at: str | None = None
    message_success_notified_at: str | None = None
    message_failure_notified_at: str | None = None
    message_notify_error: str | None = None
    created_at: str
    updated_at: str


class SapPushDispatchOut(BaseModel):
    queued: bool
    dispatch_mode: Literal["dramatiq", "background", "eager"]
    send_enabled: bool
    message: str
    log: SapPushLogOut
    logs: list[SapPushLogOut] = []

File: apps/web/src/api/recordings.ts
import { api, type PaginatedResponse } from './client'

export type Recording = {
  id: string
  visit_id: string | null
  linked_visit_ids: string[]
  linked_visits: Array<{
    id: string
    external_visit_order_no: string | null
    external_visit_order_seg: string | null
    customer_name: string | null
    is_primary: boolean
  }>
  visit_status: string | null
  staff_id: string | null
  staff_name: string | null
  staff_badge_id: string | null
  staff_role: string | null
  customer_name: string | null
  device_id: string | null
  device_code: string | null
  file_name: string
  file_path?: string
  file_size: number
  duration_seconds: number | null
  status: string
  split_parent_recording_id?: string | null
  split_part_index?: number | null
  split_at_ms?: number | null
  has_transcript: boolean
  created_at: string
}

export type MatchEvidence = {
  type: string
  label: string
  detail: string
  strength: string
}

export type VisitOrderMatchLineItem = {
  fzdh: string | null
  dzseg: string | null
  triage_staff_code: string | null
  triage_staff_name: string | null
  triage_time: string | null
  consult_time: string | null
  triage_status_text: string | null
  deal_status_text: string | null
  consult_project: string | null
  note_summary: string | null
}

export type VisitOrderMatchCandidate = {
  visit_order_id: string
  local_visit_id: string | null
  associated_local_visit_ids: string[]
  companion_visit_order_refs: string[]
  companion_customer_codes: string[]
  dzdh: string
  dzseg: string | null
  customer_name: string | null
  customer_code: string | null
  customer_type_code: string | null
  customer_type_label: string | null
  visit_date: string | null
  advisor_code: string | null
  fzuer: string | null
  fzuer_long: string | null
  triage_time: string | null
  confidence: number
  decision: string
  method: string
  reasons: string[]
  excluded_reasons: string[]
  identity_conflicts: string[]
  manual_review_required: boolean
  manual_review_reason: string | null
  evidence: MatchEvidence[]
  merged_segments: string[]
  merged_line_items: VisitOrderMatchLineItem[]
  linked_recording_count: number
  linked_recording_names: string[]
}

export type RecordingVisitOrderMatch = {
  recording_id: string
  file_name: string
  record_date: string | null
  advisor_code: string | null
  customer_code: string | null
  customer_name: string | null
  linked_visit_id: string | null
  linked_visit_ids: string[]
  linked_visit_order_refs: string[]
  linked_visit_order_no: string | null
  linked_visit_order_seg: string | null
  auto_applied: boolean
  identity_conflicts: string[]
  manual_review_required: boolean
  manual_review_reason: string | null
  summary: string
  analyzed_at: string
  candidates: VisitOrderMatchCandidate[]
}

export type AnalysisTaskRef = {
  id: string
  file_name: string
  status: 'pending' | 'running' | 'done' | 'failed'
  progress: number
  error_message: string | null
  duration_ms: number | null
  segment_count: number | null
  overall_score: number | null
  created_at: string
  updated_at: string
  completed_at: string | null
}

export type RecordingUpdatePayload = {
  visit_id?: string | null
  linked_visit_ids?: string[] | null
  staff_id?: string | null
  device_id?: string | null
  status?: string | null
  duration_seconds?: number | null
}

export type RecordingAnalysisTask = AnalysisTaskRef & {
  result: Record<string, unknown> | null
}

export type RecordingCustomerSegment = {
  id: string
  segment_index: number
  label: string
  begin_ms: number
  end_ms: number
  summary: string
  utterance_count: number
  status: string
  mapped_visit_id: string | null
}

export type RecordingVisitAnalysis = {
  id: string
  recording_id: string
  visit_id: string
  visit_order_no: string | null
  visit_order_seg: string | null
  customer_name: string | null
  customer_code: string | null
  customer_segment_id: string | null
  mapping_status: string
  analysis_status: string
  analysis_task_id: string | null
  analysis_error: string | null
  confirmed_by: string | null
  confirmed_at: string | null
  sap_ready_at: string | null
  sap_push_log_id: string | null
}

export type RecordingMultiCustomerReview = {
  recording_id: string
  required: boolean
  linked_visit_count: number
  status: 'not_required' | 'pending_mapping' | 'analyzing' | 'ready' | 'failed'
  message: string
  segments: RecordingCustomerSegment[]
  visit_analyses: RecordingVisitAnalysis[]
}

export type RecordingMediaSource = {
  url: string
  file_name: string
  media_type?: string | null
}

export const RECORDING_STATUS_MAP: Record<string, { label: string; color: string }> = {
  uploaded: { label: '待转写', color: 'default' },
  transcribing: { label: '转写中', color: 'processing' },
  transcribed: { label: '已转写', color: 'cyan' },
  analyzing: { label: '分析中', color: 'blue' },
  analyzed: { label: '分析完成', color: 'success' },
  failed: { label: '处理失败', color: 'error' },
}

export const STAFF_ROLE_MAP: Record<string, string> = {
  consultant: '咨询师',
  doctor: '医生',
  super_admin: '超级管理员',
  system_admin: '系统管理员',
  hospital_admin: '机构管理员',
  admin: '系统管理员',
  manager: '机构管理员',
  staff: '普通员工',
}

export const fetchRecordings = (params?: {
  visit_id?: string
  hospital_code?: string
  staff_id?: string
  status?: string
  keyword?: string
  customer_keyword?: string
  badge_id?: string
  role?: string
  has_visit?: boolean
  date_from?: string
  date_to?: string
  page?: number
  page_size?: number
  fast_page?: boolean
}) => {
  const sp = new URLSearchParams()
  if (params?.visit_id) sp.set('visit_id', params.visit_id)
  if (params?.hospital_code) sp.set('hospital_code', params.hospital_code)
  if (params?.staff_id) sp.set('staff_id', params.staff_id)
  if (params?.status) sp.set('status', params.status)
  if (params?.keyword) sp.set('keyword', params.keyword)
  if (params?.customer_keyword) sp.set('customer_keyword', params.customer_keyword)
  if (params?.badge_id) sp.set('badge_id', params.badge_id)
  if (params?.role) sp.set('role', params.role)
  if (params?.has_visit !== undefined) sp.set('has_visit', String(params.has_visit))
  if (params?.date_from) sp.set('date_from', params.date_from)
  if (params?.date_to) sp.set('date_to', params.date_to)
  if (params?.page) sp.set('page', String(params.page))
  if (params?.page_size) sp.set('page_size', String(params.page_size))
  if (params?.fast_page !== undefined) sp.set('fast_page', String(params.fast_page))
  const qs = sp.toString()
  return api.get(`recordings${qs ? `?${qs}` : ''}`).json<PaginatedResponse<Recording>>()
}

export const fetchRecording = (id: string) => api.get(`recordings/${id}`).json<Recording>()

export const fetchRecordingMediaSource = (id: string) =>
  api.get(`recordings/${id}/media-url`).json<RecordingMediaSource>()

export const fetchRecordingMediaBlob = (id: string) => api.get(`recordings/${id}/media`).blob()

export const uploadRecording = (
  file: File,
  opts?: { visit_id?: string; staff_id?: string; device_id?: string },
) => {
  const sp = new URLSearchParams()
  if (opts?.visit_id) sp.set('visit_id', opts.visit_id)
  if (opts?.staff_id) sp.set('staff_id', opts.staff_id)
  if (opts?.device_id) sp.set('device_id', opts.device_id)
  const qs = sp.toString()
  const formData = new FormData()
  formData.append('file', file)
  return api
    .post(`recordings/upload${qs ? `?${qs}` : ''}`, { body: formData, timeout: false })
    .json<Recording>()
}

export const updateRecording = (id: string, data: RecordingUpdatePayload) =>
  api.put(`recordings/${id}`, { json: data }).json<Recording>()

export type RecordingVisitOrderLocalVisit = {
  visit_id: string
  visit_order_id: string
  dzdh: string | null
  dzseg: string | null
}

export const ensureRecordingVisitOrderLocalVisit = (id: string, visitOrderId: string) =>
  api
    .post(`recordings/${id}/visit-order-local-visit`, { json: { visit_order_id: visitOrderId } })
    .json<RecordingVisitOrderLocalVisit>()

export const deleteRecording = (id: string) => api.delete(`recordings/${id}`)

export const analyzeRecording = (id: string) => api.post(`recordings/${id}/analyze`).json<AnalysisTaskRef>()

export type RecordingSplitResult = {
  original_recording_id: string
  split_at_ms: number
  message: string
  parts: Array<{
    part_index: number
    archive_item_id: string | null
    recording: Recording
  }>
}

export const splitRecording = (
  id: string,
  data: { split_at_seconds?: number; split_at_ms?: number; confirm: boolean },
) => api.post(`recordings/${id}/split`, { json: data, timeout: false }).json<RecordingSplitResult>()

export const fetchRecordingAnalysis = (id: string) =>
  api.get(`recordings/${id}/analysis`).json<RecordingAnalysisTask | null>()

export const fetchRecordingMultiCustomerReview = (id: string) =>
  api.get(`recordings/${id}/multi-customer-review`).json<RecordingMultiCustomerReview>()

export const confirmRecordingMultiCustomerReview = (
  id: string,
  mappings: Array<{ visit_id: string; customer_segment_id: string }>,
) => api.post(`recordings/${id}/multi-customer-review/confirm`, { json: { mappings } }).json<RecordingMultiCustomerReview>()

export const resetRecordingMultiCustomerReview = (id: string) =>
  api.post(`recordings/${id}/multi-customer-review/reset`).json<RecordingMultiCustomerReview>()

export const fetchRecordingVisitOrderMatch = (id: string, applyAuto = true, useLlm = true) => {
  const sp = new URLSearchParams()
  sp.set('apply_auto', String(applyAuto))
  sp.set('use_llm', String(useLlm))
  return api
    .get(`recordings/${id}/visit-order-match?${sp.toString()}`, { timeout: 30000 })
    .json<RecordingVisitOrderMatch>()
}

export type SapConsultationPayload = {
  text: string
  user: string
  zxxx: Record<string, string>
  TAB_SYZ: Array<Record<string, string>>
}

export type SapPushTarget = {
  visit_id: string | null
  visit_order_no: string
  visit_order_seg: string | null
  customer_name: string
  customer_code: string
  advisor_name: string
  indication_count: number
  recording_count: number
  is_primary: boolean
}

export type SapPushResult = {
  recording_id: string
  visit_order_no: string
  visit_order_seg: string | null
  customer_name: string
  customer_code: string
  advisor_name: string
  indication_count: number
  recording_count: number
  target_count: number
  targets: SapPushTarget[]
  payloads: SapConsultationPayload[]
}

export const pushRecordingToSap = (id: string) =>
  api.post(`recordings/${id}/push-sap`).json<SapPushResult>()

export type SapPushDispatchRequest = {
  trigger_mode?: 'manual' | 'auto_bind' | 'scheduled'
  async_dispatch?: boolean
  target_visit_id?: string | null
}

export type SapPushAttempt = {
  request_index: number
  success: boolean
  http_status_code: number | null
  gateway_code: number | string | null
  business_status: string | null
  business_message: string | null
  response_body: unknown
}

export type SapPushLog = {
  id: string
  recording_id: string | null
  recording_file_name: string | null
  recording_created_at: string | null
  visit_id: string | null
  visit_order_no: string | null
  visit_order_seg: string | null
  customer_name: string | null
  customer_code: string | null
  advisor_name: string | null
  trigger_mode: string
  status: string
  send_enabled: boolean
  initiated_by: string | null
  request_url: string | null
  trace_id: string | null
  request_payloads: Array<Record<string, unknown>>
  gateway_requests: Array<Record<string, unknown>>
  response_items: SapPushAttempt[]
  http_status_code: number | null
  business_status: string | null
  business_message: string | null
  error_message: string | null
  sent_at: string | null
  created_at: string
  updated_at: string
}

export type SapPushDispatchResult = {
  queued: boolean
  dispatch_mode: 'dramatiq' | 'background' | 'eager'
  send_enabled: boolean
  message: string
  log: SapPushLog
  logs: SapPushLog[]
}

export const dispatchRecordingToSap = (id: string, data?: SapPushDispatchRequest) =>
  api.post(`recordings/${id}/push-sap/dispatch`, { json: data ?? {} }).json<SapPushDispatchResult>()

export const fetchRecordingSapPushLogs = (id: string) =>
  api.get(`recordings/${id}/push-sap/logs`).json<SapPushLog[]>()

export type DailyVisitOrderItem = {
  id: string
  dzdh: string
  dzseg: string | null
  local_visit_id: string | null
  detail_local_visit_id: string | null
  associated_local_visit_ids: string[]
  companion_local_visit_ids: string[]
  companion_visit_order_refs: string[]
  companion_customer_codes: string[]
  ninam: string | null
  kunr: string | null
  customer_type_code: string | null
  customer_type_label: string | null
  sjrq: string | null
  fzsj: string | null
  fzuer: string | null
  fzuer_long: string | null
  advxc_long: string | null
  jcsta_txt: string | null
  remark_dz: string | null
  linked_recording_names: string[]
}

export type DailyVisitOrdersResponse = {
  items: DailyVisitOrderItem[]
  recording_date: string | null
  total: number
  scope_mode: 'self' | 'org'
  keyword: string
}

export const fetchDailyVisitOrdersForRecording = (
  recordingId: string,
  params?: {
    scope_mode?: 'self' | 'org'
    keyword?: string
  },
) => {
  const sp = new URLSearchParams()
  if (params?.scope_mode) sp.set('scope_mode', params.scope_mode)
  if (params?.keyword) sp.set('keyword', params.keyword)
  const qs = sp.toString()
  return api.get(`visit-orders/daily-for-recording/${recordingId}${qs ? `?${qs}` : ''}`).json<DailyVisitOrdersResponse>()
}
